tests/gem5/stats-reset/configs/x86-ubuntu-run-end-reset.py

```diff
@@ -0,0 +1,195 @@
+reset, we check their values against stats from a simulation with the
+same configuration and workload that runs for 15 billion ticks without
+resetting. We skip over stats that have values of 0 in the run that doesn't
+reset, as we cannot tell whether the stat has reset or not.
+
+"""
+
+import sys
+
+import m5
+
+from gem5.coherence_protocol import CoherenceProtocol
+from gem5.components.boards.x86_board import X86Board
+from gem5.components.processors.cpu_types import CPUTypes
+from gem5.components.processors.simple_switchable_processor import (
+    SimpleSwitchableProcessor,
+)
+from gem5.isas import ISA
+from gem5.resources.resource import obtain_resource
+from gem5.simulate.exit_event import ExitEvent
+from gem5.simulate.simulator import Simulator
+from gem5.utils.requires import requires
+
+requires(
+    isa_required=ISA.X86,
+)
+
+from gem5.components.cachehierarchies.classic.private_l1_shared_l2_cache_hierarchy import (
+    PrivateL1SharedL2CacheHierarchy,
+)
+
+cache_hierarchy = PrivateL1SharedL2CacheHierarchy(
+    l1d_size="64KiB", l1i_size="64KiB", l2_size="1MiB"
+)
+
+# Main memory
+from gem5.components.memory import DualChannelDDR4_2400
+
+memory = DualChannelDDR4_2400(size="3GiB")
+
+processor = SimpleSwitchableProcessor(
+    starting_core_type=CPUTypes.ATOMIC,
+    switch_core_type=CPUTypes.TIMING,
+    isa=ISA.X86,
+    num_cores=2,
+)
+
+board = X86Board(
+    clk_freq="3GHz",
+    processor=processor,
+    memory=memory,
+    cache_hierarchy=cache_hierarchy,
+)
+
+workload = obtain_resource("x86-ubuntu-24.04-boot-with-systemd")
+board.set_workload(workload)
+
+
+def exit_event_handler():
+    print("First exit: kernel booted")
+    processor.switch()
+    yield False  # gem5 is now executing systemd startup
+    print("Second exit: Started `after_boot.sh` script")
+    # m5.stats.dump()
+    yield False  # gem5 is now executing the `after_boot.sh` script
+    print("Third exit: Finished `after_boot.sh` script")
+    # The after_boot.sh script will run a script if it is passed via
+    # m5 readfile. This is the last exit event before the simulation exits.
+    yield True
+
+
+def max_tick_handler_end_reset():
+    m5.stats.reset()
+    m5.stats.dump()
+    yield True
+
+
+simulator = Simulator(
+    board=board,
+    on_exit_event={
+        # Here we want override the default behavior for the first m5 exit
+        # exit event.
+        ExitEvent.EXIT: exit_event_handler(),
+        ExitEvent.MAX_TICK: max_tick_handler_end_reset(),
+    },
+)
+
+simulator.set_max_ticks(15000000000)  # 15,000,000,000 ticks
+simulator.run()
+
+end_reset_dict = {}
+no_reset_dict = {}
+
+
+def read_stats_files(filepath, stats_dict):
+    with open(filepath) as stats:
+        for line in stats:
+            tmp = line.split()
+            if len(tmp) > 1:
+                stats_dict[tmp[0]] = tmp[1]
+    stats_dict.pop("----------", None)
+
+
+read_stats_files("./gem5/stats-reset/base-case-stats.txt", no_reset_dict)
```

Comment:
I think you forgot to add this to the PR?

Reply:
That stats file should be automatically generated when the first test case, which uses `x86-ubuntu-run-base.py`, is run. The first test case isn't actually a test; its sole purpose is to generate a stats file that can be used as a point of comparison for the second and third test cases. 

I set it up like this because I thought it would be more efficient than having to update the reference stats file every time a change to gem5 caused the output of the stats file to change.

Reply:
i'll give this a full review later but just to say here: you can't guarantee order of tests in testlib or declare dependencies between tests in testlib. For example, if you can just run the 2nd test here alone and it wouldn't work without running the first.

I'm guessing having this file submitted to the repo would be a bad idea as these stats may change a lot over time and break the comparison. 

Do you think what we want to do here is run two gem5 simulations from one config script. Like MultSim but just on a single thread with a guaranteed order... I wonder if such a thing is possible? We  an discuss this later but if you have time try looking into it.

Reply:
i've been thinking about this. From my understanding you want:

1. A stat file dumped at X ticks without resetting.
2. A stat file dumped at X ticks after resetting.

In which case I believe this can all be done from one simulation.

why don't you just have one exit at X ticks then in the code handling the exit you just dump the stats, move those stats somewhere safe, reset the stats and dump them again.


it wouldn't surprise me if there is some good reason you can't do this so please tell me, me if I'm wrong.

Reply:
I think that will work for the first test, but it may cause issues for the second test, which compares the results of one simulation that runs without resetting, and another that resets in the middle. Getting both sets of stats for the second test in the same simulation may cause some problems, but I can try it.

Reply:
Update: This will not work for either test. For the first test, m5.stats.dump() does not allow for multiple dumps in the same tick. For the second test, different regions of simulation (ie 0 to 15 billion ticks, then 15 to 30 billion ticks) result in different statistics, so the results can't be compared.